### ISSUE TITLE
flash_attn_with_kvcache returns incorrect output when seq_len_q > 1 AND cache_seqlens > 0 (XPU prefill path)

### ISSUE THREAD

**Issue by airMeng** [sgl-project/sgl-kernel-xpu#196]
### Summary

`flash_attn_with_kvcache` in `sgl-kernel-xpu@ab0f877` produces output that disagrees with an fp32 CPU reference by **~2.4 bf16 units max** (vs the ~1.0 bf16 noise floor) when called with `seq_len_q > 1` and `cache_seqlens > 0` (XPU prefill path). This leads to real-world model failures: e.g., Gemma-4-31B-it on tp=4 greedy-decodes the prompt "The capital of France is" to `e` (repetition collapse) instead of a coherent continuation. Triton backend on the same XPU does not have this bug. 

#### Environment
- `sgl-kernel-xpu` @ `ab0f877` (Top K Renorm Probs Kernel #182)
- `sglang` @ `main` + minor debug hooks (path unaffected)
- PyTorch 2.11.0+xpu, Python 3.12
- Hardware: Intel XPU, oneAPI 2025.3.2
- `SGLANG_USE_SGL_XPU=1`

#### Minimal Reproducer
A concise script (see below) reproduces the bug, comparing batched extend (seq_q=3, cache=4) vs. iterative (seq_q=1, cache grows 4->5->6) and vs an fp32 CPU reference. The bug appears as significant output divergence when both `seq_len_q > 1` and `cache_seqlens > 0`.

**[Full code available in the referenced markdown file for copy/paste and context.]**

#### Expected Output:

```
B iterative vs fp32-CPU      max=9.6580e-01  mean=1.0661e-01   <- ~bf16 noise
A batched   vs fp32-CPU      max=2.4406e+00  mean=3.3911e-01   <- 2.5x above noise
A batched   vs B iterative   max=2.5156e+00  mean=3.1493e-01   <- bug visible
```

#### Stronger cross-backend test
The pure-triton implementation in SGLang (`extend_attention_fwd`) [[link](https://github.com/sgl-project/sgl-kernel-xpu/issues/196#issuecomment-4427483303)] matches CPU within bf16 rounding, but intel_xpu diverges by ~2.44 bf16 units, indicating the bug is isolated to the XPU prefill kernel.

#### Trigger Conditions
| seq_len_q | cache_seqlens | Output |  
|---|---|---|
| 1 | 0 | < 0.01 — bf16 noise (decode, empty cache) |
| 1 | > 0 | 0.00 — bf16 identity (decode, dispatched decode path) |
| >1 | 0 | < 0.01 — bf16 noise (fresh prefill) |
| **>1** | **>0** | **6 – 17 bf16 units** per rank — bug |

The problematic path is "chunked prefill second chunk".

#### Narrowing and Call-site
- Inputs to the kernel are bit-identical between XPU/triton references.
- SGLang call-site: `sglang/python/sglang/srt/layers/attention/xpu_backend.py` (see markdown reference).

#### Workaround
Running without chunked prefill and radix cache disables the broken path (`--chunked-prefill-size -1 --disable-radix-cache`), restoring accuracy at some throughput cost.

#### Attempts/Falsifications
- Kernel patch attempts (see markdown file) did not resolve the issue and regressed other paths, suggesting the root cause lies deeper.
- Candidates not ruled out: tile blocking across cache-new boundary, first-tile handling, pack_gqa path.

### Asks
1. **Root-cause** the divergence inside `prefill::mha_fwd` (see hypotheses in markdown).
2. **Extend kernel tests** to explicitly cover `(seq_len_q > 1, cache_seqlens > 0, GQA ratio > 1, bfloat16)` vs. a CPU-reference, beyond current unit tests.
3. **Verify** any fix does not regress the single-step and decode paths by running `verify_workaround.py` pre/post.

**Comment by airMeng** [comment]
```python
"""Three-way reproducer: intel_xpu vs triton vs fp32-CPU on the same inputs.

Same Q/K/V/cache as reproduce_kvcache_bug.py (seq_q=3, cache=4), but also
runs triton's extend_attention_fwd. If triton agrees with CPU while
intel_xpu disagrees, the bug is strictly localized to the XPU kernel and
not a shared numerical issue.

Run:   /root/miniforge3/envs/sgl-xpu-d/bin/python reproduce_kvcache_bug_vs_triton.py
"""
from __future__ import annotations

import os

os.environ.setdefault("SGLANG_USE_SGL_XPU", "1")

import torch
from sgl_kernel.flash_attn import flash_attn_with_kvcache
from sglang.srt.layers.attention.triton_ops.extend_attention import extend_attention_fwd

DEVICE = "xpu:0"
DTYPE = torch.bfloat16

NUM_HEADS_Q = 8
NUM_HEADS_KV = 4
HEAD_DIM = 256
PAGE_SIZE = 64
CACHE_LEN = 4
NEW_LEN = 3
TOTAL_LEN = CACHE_LEN + NEW_LEN

SOFTMAX_SCALE = 1.0 / (HEAD_DIM ** 0.5)


def _zero_pad_to_page(t: torch.Tensor, page_size: int) -> torch.Tensor:
    n, h, d = t.shape
    out = torch.zeros(1, page_size, h, d, dtype=t.dtype, device=t.device)
    out[0, :n] = t
    return out


def run_intel_xpu(q, k_all, v_all):
    """intel_xpu flash_attn_with_kvcache path (the one suspected of being buggy)."""
    k_cache = _zero_pad_to_page(k_all.clone(), PAGE_SIZE)
    v_cache = _zero_pad_to_page(v_all.clone(), PAGE_SIZE)
    out = flash_attn_with_kvcache(
        q=q,
        k_cache=k_cache,
        v_cache=v_cache,
        page_table=torch.tensor([[0]], dtype=torch.int32, device=DEVICE),
        cache_seqlens=torch.tensor([CACHE_LEN], dtype=torch.int32, device=DEVICE),
        cu_seqlens_q=torch.tensor([0, NEW_LEN], dtype=torch.int32, device=DEVICE),
        cu_seqlens_k_new=torch.tensor([0, NEW_LEN], dtype=torch.int32, device=DEVICE),
        max_seqlen_q=NEW_LEN,
        softmax_scale=SOFTMAX_SCALE,
        causal=True,
    )
    return out.reshape(NEW_LEN, NUM_HEADS_Q, HEAD_DIM)


def run_triton(q, k_all, v_all):
    """triton extend_attention_fwd path.

    ABI:
      q_extend : (total_new_q, H_q, D)  = new-query tokens
      k_extend : (total_new_q, H_kv, D) = new keys (will be written into
                                          the cache at positions prefix..prefix+new)
      v_extend : (total_new_q, H_kv, D) = new values
      k_buffer, v_buffer : (cache_capacity, H_kv, D) flat  — prior cached entries
                           live at indices [0 : cache_len)
      qo_indptr : [0, NEW_LEN]          — cumulative new-query-lens
      kv_indptr : [0, CACHE_LEN]        — cumulative PRIOR-cached-lens
      kv_indices: [0, 1, ..., CACHE_LEN-1] — flat indices into k_buffer/v_buffer
    """
    k_extend = k_all[CACHE_LEN:].contiguous()  # (NEW_LEN, H_kv, D) = new K
    v_extend = v_all[CACHE_LEN:].contiguous()  # (NEW_LEN, H_kv, D) = new V

    # k_buffer holds only the prior cache entries
    k_buffer = k_all[:CACHE_LEN].contiguous()  # (CACHE_LEN, H_kv, D)
    v_buffer = v_all[:CACHE_LEN].contiguous()

    o = torch.empty(NEW_LEN, NUM_HEADS_Q, HEAD_DIM, dtype=DTYPE, device=DEVICE)

    qo_indptr = torch.tensor([0, NEW_LEN], dtype=torch.int32, device=DEVICE)
    kv_indptr = torch.tensor([0, CACHE_LEN], dtype=torch.int32, device=DEVICE)
    kv_indices = torch.arange(CACHE_LEN, dtype=torch.int64, device=DEVICE)

    extend_attention_fwd(
        q_extend=q.contiguous(),
        k_extend=k_extend,
        v_extend=v_extend,
        o_extend=o,
        k_buffer=k_buffer,
        v_buffer=v_buffer,
        qo_indptr=qo_indptr,
        kv_indptr=kv_indptr,
        kv_indices=kv_indices,
        custom_mask=None,
        is_causal=True,
        mask_indptr=None,
        max_len_extend=NEW_LEN,
        k_scale=1.0,
        v_scale=1.0,
        sm_scale=SOFTMAX_SCALE,
    )
    return o


def reference_cpu(q, k_all, v_all):
    """fp32 dense causal attention on CPU."""
    q_cpu = q.float().cpu()
    k_cpu = k_all.float().cpu()
    v_cpu = v_all.float().cpu()

    rep = NUM_HEADS_Q // NUM_HEADS_KV
    k_rep = k_cpu.repeat_interleave(rep, dim=1)
    v_rep = v_cpu.repeat_interleave(rep, dim=1)

    out = torch.zeros(NEW_LEN, NUM_HEADS_Q, HEAD_DIM, dtype=torch.float32)
    for i in range(NEW_LEN):
        visible = CACHE_LEN + i + 1
        scores = torch.einsum("hd,khd->hk", q_cpu[i], k_rep[:visible]) * SOFTMAX_SCALE
        probs = torch.softmax(scores, dim=-1)
        out[i] = torch.einsum("hk,khd->hd", probs, v_rep[:visible])
    return out


def main():
    torch.manual_seed(0)

    q = torch.randn(NEW_LEN, NUM_HEADS_Q, HEAD_DIM, dtype=DTYPE, device=DEVICE)
    k_all = torch.randn(TOTAL_LEN, NUM_HEADS_KV, HEAD_DIM, dtype=DTYPE, device=DEVICE)
    v_all = torch.randn(TOTAL_LEN, NUM_HEADS_KV, HEAD_DIM, dtype=DTYPE, device=DEVICE)

    out_intel  = run_intel_xpu(q, k_all, v_all)
    out_triton = run_triton(q, k_all, v_all)
    ref        = reference_cpu(q, k_all, v_all)

    def _diff(a, b, label):
        d = (a.float().cpu() - b.float().cpu()).abs()
        print(f"  {label:<38} max={d.max().item():.4e}  mean={d.mean().item():.4e}")

    print(f"Shape: seq_q={NEW_LEN}  cache_seqlens={CACHE_LEN}  total_k={TOTAL_LEN}")
    print(f"  heads_q={NUM_HEADS_Q}  heads_kv={NUM_HEADS_KV}  head_dim={HEAD_DIM}  dtype={DTYPE}")
    print()

    _diff(out_triton, ref,        "triton                vs fp32-CPU")
    _diff(out_intel,  ref,        "intel_xpu             vs fp32-CPU")
    _diff(out_intel,  out_triton, "intel_xpu             vs triton")

    max_intel_vs_cpu  = (out_intel.float().cpu() - ref).abs().max().item()
    max_triton_vs_cpu = (out_triton.float().cpu() - ref).abs().max().item()
    print()
    if max_triton_vs_cpu < 1.5 and max_intel_vs_cpu > 1.5:
        print("[ISOLATED] triton agrees with fp32-CPU (within bf16 noise).")
        print("           intel_xpu disagrees by >1.5 bf16 units.")
        print("           Bug is strictly inside flash_attn_with_kvcache (XPU kernel).")
    elif max_triton_vs_cpu < 1.5 and max_intel_vs_cpu < 1.5:
        print("[BOTH PASS] triton and intel_xpu both within bf16 noise of CPU.")
    else:
        print("[UNEXPECTED] triton also diverges from CPU.  Investigate.")


if __name__ == "__main__":
    main()
```

**Comment by Valentine233** [comment]
Expected to be fixed by https://github.com/sgl-project/sgl-kernel-xpu/pull/191.

**Comment by Valentine233** [comment]
Validated to be fixed by https://github.com/sgl-project/sglang/pull/23280.

**Comment by Valentine233** [comment]
Fixed.
